jni/include/knn_extension/faiss/utils/Heap.h

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <climits>
+#include <cmath>
+#include <cstring>
+
+#include <stdint.h>
+#include <cassert>
+#include <cstdio>
+
+#include <limits>
+#include <unordered_map>
+#include <faiss/utils/ordered_key_value.h>
+
+// Collection of heap operations with parent id to dedupe
+namespace os_faiss {
+
+/**
+ * From start_index, it compare its value with parent node's and swap if needed.
+ * Continue until either there is no swap or it reaches the top node.
+ *
+ * @param bh_val        binary heap storing values
+ * @param bh_ids        binary heap storing parent ids
+ * @param val           new value to add
+ * @param id            new id to add
+ * @parent_id_to_id     parent doc id to id mapping data, see MultiVectorResultCollector.h
+ * @parent_id_to_index  parent doc id to index mapping data, see MultiVectorResultCollector.h
+ * @parent_id           parent id of given id
+ * @start_index         an index to start up-heap from in the binary heap(bh_val, and bh_ids)
+ */
+template <class C>
+static inline void up_heap(
+        typename C::T* bh_val,
+        typename C::TI* bh_ids,
+        typename C::T val,
+        typename C::TI id,
+        std::unordered_map<typename C::TI, typename C::TI>* parent_id_to_id,
+        std::unordered_map<typename C::TI, size_t>* parent_id_to_index,
+        typename C::TI parent_id,
+        size_t start_index) {
+    bh_val--; /* Use 1-based indexing for easier node->child translation */
+    bh_ids--;
+    size_t i = start_index + 1, i_father;
+
+    while (i > 1) {
```

Comment:
does this make sense if start_index=0? Would that mean we are just going to set the id as the first one?

Reply:
Yes. That is correct.